fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/MaterializedViewRewriter.java

```diff
@@ -2379,17 +2495,50 @@ private ScalarOperator getCompensationEqualPredicate(EquivalenceClasses sourceEq
                 // it means that the targetEc do not have the corresponding mapping ec
                 // we should all equality predicates between each column in ec into compensation
                 Iterator<ColumnRefOperator> it = sourceEquivalenceClassesList.get(i).iterator();
-                ScalarOperator first = it.next();
+                ColumnRefOperator first = it.next();
+
+                // ignore if first is redundant
+                if (sourceEquivalenceClasses.containsRedundantKey(first)) {
+                    sourceEquivalenceClasses.deleteRedundantKey(first);
+                    continue;
+                }
+
                 while (it.hasNext()) {
-                    ScalarOperator equalPredicate = BinaryPredicateOperator.eq(first, it.next());
+                    ColumnRefOperator next = it.next();
+                    ScalarOperator equalPredicate = BinaryPredicateOperator.eq(first, next);
                     compensation = Utils.compoundAnd(compensation, equalPredicate);
                 }
             } else {
                 // remove columns exists in target and add remain equality predicate in source into compensation
                 for (int j : mapping.get(i)) {
-                    Set<ScalarOperator> difference = Sets.newHashSet(sourceEquivalenceClassesList.get(i));
+                    Set<ColumnRefOperator> difference = Sets.newHashSet(sourceEquivalenceClassesList.get(i));
                     Set<ColumnRefOperator> targetColumnRefs = targetEquivalenceClassesList.get(j);
                     difference.removeAll(targetColumnRefs);
+                    if (difference.size() == 0) {
+                        continue;
+                    }
+
+                    {
+                        // remove redundant from equivalence
+                        Iterator<ColumnRefOperator> it = difference.iterator();
```

Comment:
what if a = b and a = 1 and c =1, do this work ok?

Reply:
yep. redundant keys only record the keys which is not in ec.

```
    public boolean addRedundantEquivalence(ColumnRefOperator left, ColumnRefOperator right) {
        boolean isContainsLeft = columnToEquivalenceClass.containsKey(left);
        boolean isContainsRight = columnToEquivalenceClass.containsKey(right);
        boolean isAddIntoRedundant = false;
        if (!isContainsLeft) {
            redundantColumnToEquivalenceClass.computeIfAbsent(left, x -> Sets.newHashSet(left));
            if (!isContainsRight) {
                redundantColumnToEquivalenceClass.get(left).add(right);
            }
            isAddIntoRedundant = true;
        }
        if (!isContainsRight) {
            redundantColumnToEquivalenceClass.computeIfAbsent(right, x -> Sets.newHashSet(right));
            if (!isContainsLeft) {
                redundantColumnToEquivalenceClass.get(right).add(left);
            }
            isAddIntoRedundant = true;
        }
        return isAddIntoRedundant;
    }
```